sympy/polys/matrices/ddm.py

```diff
@@ -331,3 +348,17 @@ def charpoly(a):
 
 
 from .sdm import SDM
+
+
+def ddm_transpose(M):
+
+    nrows, ncols = M.shape
+    MT = []
+
+    for j in range(ncols):
+        row = []
+        for i in range(nrows):
+            row.append(M[i][j])
+        MT.append(row)
```

Comment:
A function `ddm_transpose` would be expected to operate just on a list of lists so there would be no `shape` attribute.

Also I'm pretty sure that the `zip` version is a lot faster than this.

Reply:
`zip` method doesn't really work with shape `(0, 2)`

Reply:
That case will need special handling. This function that only works with a list of lists can't handle the `(0, 2)` case at all.

You should do timings though. Matrix transpose is a performance sensitive operation and should be cheap. The `ddm_` function for `mul` uses transpose as its first step for example.

Reply:
We need the shape to get transpose of list of lists as `[]` can represent DDM of shape `(0, n)` for any n >= 0, but transpose of `[]` varies depending on shape `(0,n)`. The transpose of `[]` for shape `(0,n)` is `[[]]*n`.

Using `zip` method we get `[]` as the transpose  of `DDM([], (0, n), ZZ)`  instead of `DDM([[]]*n, (n, 0), ZZ)`.

Reply:
A very simple fix would be to simply pass the `shape` as argument to ddm_transpose like this


```python
def ddm_transpose(M, shape):
    MT = []
    nrows,ncols = shape
    
    for j in range(ncols):
        row = []
        for i in range(nrows):
            row.append(M[i][j])
        MT.append(row)

    return MT
```

Reply:
In most cases we don't really care about this distinction between the shape of empty matrices like this e.g.:
https://github.com/sympy/sympy/blob/0f807b64436da37b1ab90a595144dbbdbef27d91/sympy/polys/matrices/dense.py#L75-L79
Maybe it would be cleaner if all of the `ddm_*` functions just operated with `DDM` but I think that for this particular case it's easiest just to handle the `(0, 2)` case in `DDM.transpose` and have `ddm_transpose` return an empty list in that case.